### PR TITLE
♻️ Support img element

### PR DESCRIPTION
**Pull request by kristoferbaxter**
This PR is not complete, but adds support for `img` elements elements to the AMP runtime.

- [x] Ensure existing functionality works with `img`.
- [x] Test coverage for each item expecting `amp-img` when it receives an `img`.
- [x] Test for the addition to 'src/core/dom' following the semantics of other 'core' utilities.

Originally this PR was also going to add support for `picture`, but the scope was getting larger than a single PR really should have. I'll split it apart.

### CODE REVIEW

File: src/core/dom/propagateAttributes.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * Utility method that propagates attributes from a source element
+ * to an updateable element.
+ * If `opt_removeMissingAttrs` is true, then also removes any specified
+ * attributes that are missing on the source element from the updateable element.
+ * @param {string|!Array<string>} attributes
+ * @param {!Element} sourceElement
+ * @param {!Element} updateElement
+ * @param {boolean=} opt_removeMissingAttrs
+ * @public @final
```
Comment (rcebulko) [inline src/core/dom/propagateAttributes.js:28]: What do `@public @final` do here?

File: src/core/dom/propagateAttributes.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ */
+export function propagateAttributes(
+  attributes,
+  sourceElement,
+  updateElement,
+  opt_removeMissingAttrs
+) {
+  attributes = isArray(attributes) ? attributes : [attributes];
```
Comment (rcebulko) [inline src/core/dom/propagateAttributes.js:36]: https://github.com/ampproject/amphtml/blob/fe223daa62862438a6023fa7ab53b500d3074242/src/core/types/array.js#L41-L45

File: extensions/amp-image-lightbox/0.1/amp-image-lightbox.js [hunk 3]
```diff
@@ -49,6 +50,11 @@ const SUPPORTED_ELEMENTS_ = {
   'amp-img': true,
   'amp-anim': true,
 };
+/** @private @const {!Object<string, boolean>} */
+const SUPPORTED_NATIVE_ELEMENTS = {
```
Comment (rcebulko) [inline extensions/amp-image-lightbox/0.1/amp-image-lightbox.js:54]: Maybe consider a `Set` or `Map` for this; might be uglier though

**Review by rcebulko — COMMENTED** [review]
Love seeing class methods get turned into pure functions. Nice!
Suggestion: rename to `propagate-attributes.js` for consistency

File: src/core/dom/propagateAttributes.js (hunk 1 above)
Comment (kristoferbaxter) [reply src/core/dom/propagateAttributes.js:28]: They're unnecessary here, copy/paste issue bringing this over from AMP.BaseElement.

File: extensions/amp-image-lightbox/0.1/amp-image-lightbox.js (hunk 3 above)
Comment (kristoferbaxter) [reply extensions/amp-image-lightbox/0.1/amp-image-lightbox.js:54]: A `Set` makes sense, can add to this PR.

File: examples/image-lightbox.amp.html [hunk 4]
```diff
@@ -22,6 +22,13 @@
 <script async src="https://cdn.ampproject.org/v0.js"></script>
 </head>
 <body>
+  <!-- Standard Image Element -->
```
Comment (kristoferbaxter) [inline examples/image-lightbox.amp.html:25]: Include in this comment that `img` elements are not yet valid, but will be.

File: examples/amp-lightbox.amp.html [hunk 5]
```diff
@@ -63,6 +63,7 @@ <h2>Scrollable Lightbox</h2>
       Lorem ipsum dolor sit amet, has nisl nihil convenire et, vim at aeque inermis reprehendunt. Propriae tincidunt id nec, elit nusquam te mea, ius noster platonem in. Mea an idque minim, sit sale deleniti apeirian et. Omnium legendos tractatos cu mea. Vix in stet dolorem accusamus. Iisque rationibus consetetur in cum, quo unum nulla legere ut. Simul numquam saperet no sit.
     </p>
     <p>
+      <img src="https://lh3.googleusercontent.com/5rcQ32ml8E5ONp9f9-Rf78IofLb9QjS5_0mqsY1zEFc=w300-h200-no" width=300 height=200 loading="lazy" />
```
Comment (kristoferbaxter) [inline examples/amp-lightbox.amp.html:66]: Include a comment that img elements are not yet valid, but will be.

File: extensions/amp-auto-lightbox/0.1/amp-auto-lightbox.js [hunk 6]
```diff
@@ -109,6 +109,9 @@ const DISABLED_ANCESTORS = [
   'amp-carousel',
 ].join(',');
 
+/** @type {Set<string>} */
```
Comment (kristoferbaxter) [inline extensions/amp-auto-lightbox/0.1/amp-auto-lightbox.js:112]: Add a comment that later this will include Picture, thus is a Set.

Or, just make this a comparison to `img` for now with a comment.

File: extensions/amp-auto-lightbox/0.1/amp-auto-lightbox.js [hunk 7]
```diff
@@ -161,7 +164,7 @@ export class Criteria {
    */
   static meetsSizingCriteria(element, renderWidth, renderHeight) {
     const {naturalWidth, naturalHeight} = getMaxNaturalDimensions(
-      dev().assertElement(element.querySelector('img'))
+      dev().assertElement(element.querySelector('img') || element)
```
Comment (kristoferbaxter) [inline extensions/amp-auto-lightbox/0.1/amp-auto-lightbox.js:167]: Increase test coverage to justify that `element` is an `img` and doesn't break otherwise.

File: extensions/amp-auto-lightbox/0.1/amp-auto-lightbox.js [hunk 8]
```diff
@@ -298,7 +296,7 @@ export class Scanner {
    * @return {!Array<!Element>}
    */
   static getCandidates(root) {
-    const selector = candidateSelector('amp-img');
+    const selector = `amp-img:not([${LIGHTBOXABLE_ATTR}]):not([${VISITED_ATTR}]),img:not([${LIGHTBOXABLE_ATTR}]):not([${VISITED_ATTR}])`;
```
Comment (kristoferbaxter) [inline extensions/amp-auto-lightbox/0.1/amp-auto-lightbox.js:299]: Add some clarity to this selector, it's a little much.

File: extensions/amp-auto-lightbox/0.1/amp-auto-lightbox.js [hunk 9]
```diff
@@ -440,11 +438,16 @@ export function runCandidates(ampdoc, candidates) {
     whenLoaded(candidate).then(() => {
       return measureIntersectionNoRoot(candidate).then(
         ({boundingClientRect}) => {
-          // <amp-img> will change the img's src inline data on unlayout and
-          // remove it from DOM.
-          if (!candidate.signals().get(CommonSignals.LOAD_END)) {
+          if (NATIVE_ELEMENTS_SUPPORTED.has(candidate.tagName.toLowerCase())) {
```
Comment (kristoferbaxter) [inline extensions/amp-auto-lightbox/0.1/amp-auto-lightbox.js:441]: Makes more sense to explicitly call out `img` until `picture` support is added.

File: extensions/amp-image-lightbox/0.1/amp-image-lightbox.js [hunk 10]
```diff
@@ -39,16 +39,16 @@ import {
...
 
-/** @private @const {!Object<string, boolean>} */
-const SUPPORTED_ELEMENTS_ = {
-  'amp-img': true,
-  'amp-anim': true,
-};
+/** @private @const {!Set<string>} */
+const SUPPORTED_ELEMENTS_ = new Set(['amp-img', 'amp-anim']);
+/** @private @const {!Set<string>} */
+const SUPPORTED_NATIVE_ELEMENTS = new Set(['img']);
```
Comment (kristoferbaxter) [inline extensions/amp-image-lightbox/0.1/amp-image-lightbox.js:51]: Question here too, should this be a Set yet?

File: extensions/amp-lightbox-gallery/0.1/service/lightbox-manager-impl.js [hunk 11]
```diff
@@ -38,18 +38,14 @@ import {map} from '../../../../src/core/types/object';
 import {srcsetFromElement, srcsetFromSrc} from '../../../../src/srcset';
 import {toArray} from '../../../../src/core/types/array';
 
-const LIGHTBOX_ELIGIBLE_TAGS = {
-  'AMP-IMG': true,
-};
+const LIGHTBOX_ELIGIBLE_TAGS = new Set(['AMP-IMG']);
+const LIGHTBOX_ELGIBLE_NATIVE_ELEMENTS = new Set(['IMG']);
```
Comment (kristoferbaxter) [inline extensions/amp-lightbox-gallery/0.1/service/lightbox-manager-impl.js:42]: Curious if `IMG` can be added to the default `ELIGBLE_TAGS` list.

**Review by kristoferbaxter — COMMENTED** [review]
Self review for current status. Close!

File: extensions/amp-image-lightbox/0.1/amp-image-lightbox.js (hunk 10 above)
Comment (kristoferbaxter) [reply extensions/amp-image-lightbox/0.1/amp-image-lightbox.js]: Removed this set.

File: src/error-reporting.js [hunk 12]
```diff
@@ -715,15 +715,17 @@ export function detectJsEngineFromStack() {
...
-    // IE looks like Chrome, but includes a context for the base stack line.
-    // Explicitly, we're looking for something like:
-    // "    at Global code (https://example.com/app.js:1:200)" or
-    // "    at Anonymous function (https://example.com/app.js:1:200)"
-    // vs Chrome which has:
-    // "    at https://example.com/app.js:1:200"
-    const last = stack.split('\n').pop();
-    if (/\bat .* \(/i.test(last)) {
-      return 'IE';
+    if (!IS_ESM) {
```
Comment (kristoferbaxter) [inline src/error-reporting.js:718]: Unrelated change to pull out.

File: extensions/amp-image-lightbox/0.1/amp-image-lightbox.js [hunk 13]
```diff
@@ -869,8 +873,9 @@ class AmpImageLightbox extends AMP.BaseElement {
     this.buildLightbox_();
 
     const source = invocation.caller;
+    const tagName = source.tagName.toLowerCase();
     userAssert(
-      source && SUPPORTED_ELEMENTS_[source.tagName.toLowerCase()],
+      source && (SUPPORTED_AMP_ELEMENTS_.has(tagName) || 'img' === tagName),
```
Comment (alanorozco) [inline extensions/amp-image-lightbox/0.1/amp-image-lightbox.js:878]: nit: This allowlist is not used anywhere else, so the distinction between AMP/non-AMP elements might not be necessary. Consider adding `img` to the list instead of the special condition.

**Review by alanorozco — COMMENTED** [review]
Looks good so far.

I found the following allowlists and selectors that were not addressed. If we're not looking to address them, it might be valuable to add comments that explain the exclusion of `img`.

```
extensions/amp-image-slider/0.1/amp-image-slider.js:
  116        const child = children[i];
  117:       if (child.tagName.toLowerCase() === 'amp-img') {
  118          // First encountered = left image

extensions/amp-image-viewer/0.1/amp-image-viewer.js:
  58  const ELIGIBLE_TAGS = {
  59:   'amp-img': true,
  60    'amp-anim': true,

extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js:
  904        const srcCropEl =
  905:         closestAncestorElementBySelector(srcImg, 'amp-img') || srcImg;
  906        const targetCropEl =
  907:         closestAncestorElementBySelector(targetImg, 'amp-img') || targetImg;

extensions/amp-pan-zoom/0.1/amp-pan-zoom.js:
  51    'DIV': true,
  52:   'AMP-IMG': true,
  53    'AMP-LAYOUT': true,

extensions/amp-script/0.1/amp-script.js:
  716      // For now, only allow built-in AMP components except amp-pixel.
  717:     this.allowedTags_['amp-img'] = true;
  718      this.allowedTags_['amp-layout'] = true;

extensions/amp-story-360/0.1/amp-story-360.js:
  678    layoutCallback() {
  679:     const ampImgEl = this.element.querySelector('amp-img');
  680      // Used to update the video in animate_.
```

File: extensions/amp-auto-lightbox/0.1/amp-auto-lightbox.js [hunk 14]
```diff
@@ -440,13 +448,17 @@ export function runCandidates(ampdoc, candidates) {
     whenLoaded(candidate).then(() => {
       return measureIntersectionNoRoot(candidate).then(
         ({boundingClientRect}) => {
-          // <amp-img> will change the img's src inline data on unlayout and
-          // remove it from DOM.
-          if (!candidate.signals().get(CommonSignals.LOAD_END)) {
+          if (
+            !candidate.tagName === 'IMG' &&
```
Comment (rcebulko) [inline extensions/amp-auto-lightbox/0.1/amp-auto-lightbox.js:452]: Convention from what I've seen:
```suggestion
            !candidate.tagName.toLowerCase() === 'img' &&
```

Side note: I hate this convention/requirement. `toLowerCase` is 11 characters that probably can't even be minified anywhere, and it appears 24 times in v0.js and (if a rough grep is reliable) ~4000 times total in dist. According to MDN, it should always be capitalized for [DOM trees representing HTML documents](https://developer.mozilla.org/en-US/docs/Web/API/Element/tagName#:~:text=For%20DOM%20trees%20which%20represent%20HTML%20documents). Maybe we can just do it the way you do above?

File: extensions/amp-auto-lightbox/0.1/test/test-amp-auto-lightbox.js [hunk 15]
```diff
@@ -140,29 +140,62 @@ describes.realWin(
...
+                    <div>
+                      <img src="carnitas.png" layout="flex-item" />
+                    </div>
+                  </div>
+                `,
+                tagName: 'IMG',
+              },
             ].forEach((unwrapped) => {
-              maybeMutate(unwrapped);
+              maybeMutate(unwrapped.markup);
```
Comment (rcebulko) [inline extensions/amp-auto-lightbox/0.1/test/test-amp-auto-lightbox.js:190]: Nit suggestion: consider destructuring in args

File: extensions/amp-image-lightbox/0.1/test/test-amp-image-lightbox.js [hunk 16]
```diff
@@ -112,6 +106,52 @@ describes.realWin(
...
+
+      const viewer = container.querySelector(
+        '.i-amphtml-image-lightbox-viewer'
+      );
+      expect(viewer).to.not.equal(null);
+
+      const image = viewer.querySelector(
+        '.i-amphtml-image-lightbox-viewer-image'
+      );
+      expect(image).to.not.equal(null);
```
Comment (rcebulko) [inline extensions/amp-image-lightbox/0.1/test/test-amp-image-lightbox.js:146]: nit for meaningful syntax
```suggestion
      expect(container).to.not.be.null;

      const caption = container.querySelector(
        '.i-amphtml-image-lightbox-caption'
      );
      expect(caption).to.not.be.null;
      expect(caption).to.have.class('amp-image-lightbox-caption');

      const viewer = container.querySelector(
        '.i-amphtml-image-lightbox-viewer'
      );
      expect(viewer).to.not.be.null;

      const image = viewer.querySelector(
        '.i-amphtml-image-lightbox-viewer-image'
      );
      expect(image).to.not.be.null;
```

File: extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js [hunk 17]
```diff
@@ -275,11 +275,11 @@ export class AmpLightboxGallery extends AMP.BaseElement {
...
    */
   cloneLightboxableElement_(element) {
-    const fallback = element.getFallback();
-    const shouldCloneFallback =
-      element.classList.contains('amp-notsupported') && !!fallback;
-    if (shouldCloneFallback) {
-      element = fallback;
+    if (element.classList.contains('amp-notsupported')) {
+      const fallback = element.getFallback();
+      if (!!fallback) {
```
Comment (rcebulko) [inline extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js:280]: truthy/falsey should be fine here
```suggestion
      if (fallback) {
```

File: extensions/amp-lightbox-gallery/0.1/service/lightbox-manager-impl.js [hunk 18]
```diff
@@ -38,18 +38,13 @@ import {map} from '../../../../src/core/types/object';
...
-};
+// eslint-disable-next-line local/no-export-side-effect
+export const ELIGIBLE_TAP_TAGS = new Set(['AMP-IMG']);
 
-export const VIDEO_TAGS = {
-  'AMP-YOUTUBE': true,
-  'AMP-VIDEO': true,
-};
+// eslint-disable-next-line local/no-export-side-effect
+export const VIDEO_TAGS = new Set(['AMP-YOUTUBE', 'AMP-VIDEO']);
```
Comment (rcebulko) [inline extensions/amp-lightbox-gallery/0.1/service/lightbox-manager-impl.js:47]: I believe I was the one who suggested a set over an object here a while ago, but have since learned ways non-static modules can be problematic for compiler/bento. Since this is a 0.1 I'm guessing it's fine, but consider double-checking with Justin and/or Jake that it's okay here

File: src/core/dom/propagate-attributes.js [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+  attributes,
+  sourceElement,
+  updateElement,
+  opt_removeMissingAttrs
+) {
+  const attrs = /** @type {Array<string>} */ (
+    arrayOrSingleItemToArray(attributes)
+  );
+  for (let i = 0; i < attrs.length; i++) {
+    const attr = attrs[i];
```
Comment (rcebulko) [inline src/core/dom/propagate-attributes.js:39]: type ambiguity resolved in #34355, so this should pass typechecking without issue.
```suggestion
  for (const attr of arrayOrSingleItemToArray(attributes)) {
```

File: test/unit/test-amp-img.js [hunk 20]
```diff
@@ -406,7 +407,10 @@ describes.sandboxed('amp-img', {}, (env) => {
     el.getLayoutSize = () => ({width: 300, height: 200});
 
     const impl = new AmpImg(el);
-    const propagateAttributesSpy = sandbox.spy(impl, 'propagateAttributes');
+    const propagateAttributesSpy = sandbox.spy(
```
Comment (rcebulko) [inline test/unit/test-amp-img.js:410]: I vaguely recall Justin saying it's problematic that we use this pattern to spy on module functions, but I don't know how problematic or what the workaround would be

File: extensions/amp-image-lightbox/0.1/amp-image-lightbox.js (hunk 13 above)
Comment (kristoferbaxter) [reply extensions/amp-image-lightbox/0.1/amp-image-lightbox.js:878]: Originally did this, but moved away since it was strange to have a mix of AMP Elements and HTMLElements.

Will change though, this seems like a good thing to address.

File: extensions/amp-auto-lightbox/0.1/amp-auto-lightbox.js (hunk 14 above)
Comment (kristoferbaxter) [reply extensions/amp-auto-lightbox/0.1/amp-auto-lightbox.js:452]: I'm fine with making this change.

**Review by rebeccanthomas — APPROVED** [review]

File: extensions/amp-lightbox-gallery/0.1/service/lightbox-manager-impl.js (hunk 18 above)
Comment (kristoferbaxter) [reply extensions/amp-lightbox-gallery/0.1/service/lightbox-manager-impl.js:47]: @samouri or @jridgewell what's the issue with exporting a `Set`?

File: src/core/dom/propagate-attributes.js (hunk 19 above)
Comment (kristoferbaxter) [reply src/core/dom/propagate-attributes.js]: Applied a change, but kept the const attribution before the loop definition.

File: test/unit/test-amp-img.js (hunk 20 above)
Comment (kristoferbaxter) [reply test/unit/test-amp-img.js:410]: @jridgewell I remember that too now, what pattern should I use instead here?

File: extensions/amp-lightbox-gallery/0.1/service/lightbox-manager-impl.js (hunk 18 above)
Comment (samouri) [reply extensions/amp-lightbox-gallery/0.1/service/lightbox-manager-impl.js:47]: I was unsure whether or not module-level `Set` could be DCEd. If the compiler weren't smart enough to realize that the `Set` constructor was side-effect free, then it would be force to leave it in.

I [just confirmed](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%250Aconst%2520a%2520%253D%2520new%2520Set()%253B%250Aconst%2520b%2520%253D%2520%2522hello%2520world%2522%253B%250Aconst%2520c%2520%253D%2520%2522hello%2520world%2522%253B%250A%250Aconsole.log(c)%253B) that instantiating Set in module level can be DCEd by Closure.

**Review by rcebulko — APPROVED** [review]

File: extensions/amp-image-slider/0.1/amp-image-slider.js [hunk 21]
```diff
@@ -138,23 +138,31 @@ export class AmpImageSlider extends AMP.BaseElement {
             'Should not contain <div>s without ' +
               '"first" or "second" attributes.'
           );
+          break;
```
[the remark this answers is not in the document]
Comment (kristoferbaxter) [reply extensions/amp-image-slider/0.1/amp-image-slider.js:141]: These are error states, and the component will not work as expected when these are encountered. The break added was to short-circuit additional work in the loop.

Happy to remove as part of this PR and do separately though.

File: extensions/amp-image-slider/0.1/amp-image-slider.js [hunk 22]
```diff
@@ -708,45 +716,44 @@ export class AmpImageSlider extends AMP.BaseElement {
...
+      // SIMPLER SOL: simply always call scheduleLayout no matter what
+      const owners = Services.ownersForDoc(this.element);
+      owners.scheduleLayout(this.element, dev().assertElement(this.leftImage_));
+      owners.scheduleLayout(
+        this.element,
+        dev().assertElement(this.rightImage_)
+      );
+
+      return Promise.all([
+        dev()
```
Comment (kristoferbaxter) [inline extensions/amp-image-slider/0.1/amp-image-slider.js:742]: So far all the examples and test do not mix AMP Images and HTMLImageElements. Perhaps consider allowing this.

**Comment by kristoferbaxter** [comment]
Resolved extensions/amp-image-slider/0.1/amp-image-slider.js.

File: extensions/amp-image-viewer/0.1/amp-image-viewer.js [hunk 23]
```diff
@@ -319,9 +319,13 @@ export class AmpImageViewer extends AMP.BaseElement {
...
+      st.toggle(img, false);
       this.element.appendChild(this.image_);
-      return ampImg.getImpl().then((impl) => {
+      if (img.tagName === 'IMG') {
+        propagateAttributes(ARIA_ATTRIBUTES, img, this.image_);
+        return Promise.resolve();
+      }
+      return img.getImpl().then((impl) => {
         propagateAttributes(ARIA_ATTRIBUTES, impl.element, this.image_);
       });
```
Comment (rcebulko) [inline extensions/amp-image-viewer/0.1/amp-image-viewer.js]: For your consideration, this might make it a little clearer that the same thing is happening in both cases

```suggestion
      const imgPromise = img.tagName === 'IMG')
        ? Promise.resolve(img)
        : img.getImpl().then((impl) => impl.element);
      return imgPromise.then((imgEl) => {
        propagateAttributes(ARIA_ATTRIBUTES, imgEl, this.image_);
      });
```

File: extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js [hunk 24]
```diff
@@ -1369,6 +1369,7 @@ export class AmpLightboxGallery extends AMP.BaseElement {
     </div>`;
     const imgElement = childElementByTag(element, 'img');
 
+    console.log({thumbnailObj});
```
Comment (rcebulko) [inline extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js:1372]:
```suggestion
```

File: extensions/amp-lightbox-gallery/0.1/service/lightbox-manager-impl.js [hunk 25]
```diff
@@ -41,13 +41,13 @@ import {toArray} from '../../../../src/core/types/array';
 const LIGHTBOX_ELIGIBLE_TAGS = new Set(['AMP-IMG', 'IMG']);
 
 // eslint-disable-next-line local/no-export-side-effect
-export const ELIGIBLE_TAP_TAGS = new Set(['AMP-IMG']);
+export const ELIGIBLE_TAP_TAGS = new Set(['AMP-IMG', 'IMG']);
```
Comment (rcebulko) [inline extensions/amp-lightbox-gallery/0.1/service/lightbox-manager-impl.js:44]: Since this is pattern we're embracing, it might be worth filing an FR to update our custom `no-export-side-effect` rule to allowlist Set/Map

File: extensions/amp-lightbox-gallery/0.1/service/lightbox-manager-impl.js [hunk 26]
```diff
@@ -121,9 +121,9 @@ export class LightboxManager {
     // mapped unique ids.
     /**
      * List of lightbox elements that have already been scanned.
-     * @private {!Array<!Element>}
+     * @private {!Set<!Element>}
      */
-    this.seen_ = [];
+    this.seen_ = new Set([]);
```
Comment (rcebulko) [inline extensions/amp-lightbox-gallery/0.1/service/lightbox-manager-impl.js:126]: Super-nit:
```suggestion
    this.seen_ = new Set();
```

File: extensions/amp-lightbox-gallery/0.1/service/lightbox-manager-impl.js [hunk 27]
```diff
@@ -400,7 +402,10 @@ export class LightboxManager {
     if (element.hasAttribute('lightbox-thumbnail-id')) {
       const thumbnailId = element.getAttribute('lightbox-thumbnail-id');
       const thumbnailImage = this.ampdoc_.getElementById(thumbnailId);
-      if (thumbnailImage && thumbnailImage.tagName == 'AMP-IMG') {
+      if (
+        thumbnailImage &&
+        LIGHTBOX_ELIGIBLE_TAGS.has(thumbnailImage.tagName)
+      ) {
```
Comment (rcebulko) [inline extensions/amp-lightbox-gallery/0.1/service/lightbox-manager-impl.js:408]:
```suggestion
      if (LIGHTBOX_ELIGIBLE_TAGS.has(thumbnailImage?.tagName)) {
```

File: extensions/amp-pan-zoom/0.1/amp-pan-zoom.js [hunk 28]
```diff
@@ -46,13 +47,14 @@ const TAG = 'amp-pan-zoom';
...
 
-const ELIGIBLE_TAGS = {
-  'svg': true,
-  'DIV': true,
-  'AMP-IMG': true,
-  'AMP-LAYOUT': true,
-  'AMP-SELECTOR': true,
-};
+const ELIGIBLE_TAGS = new Set([
+  'svg',
```
Comment (rcebulko) [inline extensions/amp-pan-zoom/0.1/amp-pan-zoom.js:51]: Do you happen to know why 'svg' is the only lowercase here? I recognize it was like that beforehand

File: extensions/amp-pan-zoom/0.1/amp-pan-zoom.js [hunk 29]
```diff
@@ -173,8 +175,12 @@ export class AmpPanZoom extends AMP.BaseElement {
       '%s should have its target element as its one and only child',
       TAG
     );
+    console.log('attempt', {
+      valid: ELIGIBLE_TAGS.has(children[0].tagName),
+      child: children[0],
+    });
```
Comment (rcebulko) [inline extensions/amp-pan-zoom/0.1/amp-pan-zoom.js:181]: Just leaving these here to make sure stragglers don't survive CI :)
```suggestion
```

File: extensions/amp-pan-zoom/0.1/amp-pan-zoom.js (hunk 28 above)
Comment (kristoferbaxter) [reply extensions/amp-pan-zoom/0.1/amp-pan-zoom.js:51]: https://www.w3.org/TR/REC-DOM-Level-1/level-one-core.html#ID-745549614

File: extensions/amp-image-slider/0.1/amp-image-slider.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (kristoferbaxter) [reply extensions/amp-image-slider/0.1/amp-image-slider.js]: Removed from this PR, will followup separately.

File: extensions/amp-image-slider/0.1/amp-image-slider.js [hunk 30]
```diff
@@ -42,10 +45,11 @@ export class AmpImageSlider extends AMP.BaseElement {
     this.container_ = null;
 
     /** @private {?Element} */
-    this.leftAmpImage_ = null;
-
+    this.leftImage_ = null;
     /** @private {?Element} */
-    this.rightAmpImage_ = null;
+    this.rightImage_ = null;
+    /** @private {boolean} */
```
Comment (rcebulko) [inline extensions/amp-image-slider/0.1/amp-image-slider.js:51]: Nit: whitespace convention
```suggestion

    /** @private {boolean} */
```

File: extensions/amp-image-slider/0.1/amp-image-slider.js [hunk 31]
```diff
@@ -112,22 +112,21 @@ export class AmpImageSlider extends AMP.BaseElement {
   buildCallback() {
     const children = this.getRealChildren();
 
-    for (let i = 0; i < children.length; i++) {
-      const child = children[i];
-      if (child.tagName.toLowerCase() === 'amp-img') {
+    for (const child of children) {
```
Comment (rcebulko) [inline extensions/amp-image-slider/0.1/amp-image-slider.js:115]: Re: chat thread earlier on for..of
```suggestion
    for (let i = 0; i < children.length; i++) {
      const child = children[i];
```

File: extensions/amp-image-slider/0.1/amp-image-slider.js [hunk 32]
```diff
@@ -295,8 +296,13 @@ export class AmpImageSlider extends AMP.BaseElement {
...
    */
   checkARIA_() {
-    const leftAmpImage = dev().assertElement(this.leftAmpImage_);
-    const rightAmpImage = dev().assertElement(this.rightAmpImage_);
+    if (!this.containsAMPImages_) {
+      return;
+    }
+
+    // Only if there are AMP-IMG Elements in use should this pathway execute.
+    const leftAmpImage = dev().assertElement(this.leftImage_);
```
Comment (rcebulko) [inline extensions/amp-image-slider/0.1/amp-image-slider.js:304]: A bit odd to see two different conventions for AMP capitalization (`leftAmpImage` here or `AmpElement` vs `containsAMPImages` above). `this.containsAmpImages_` (while admittedly uglier) might align better

**Review by rcebulko — APPROVED** [review]

File: extensions/amp-image-slider/0.1/amp-image-slider.js (hunk 31 above)
Comment (kristoferbaxter) [reply extensions/amp-image-slider/0.1/amp-image-slider.js:115]: Turns out `for ... of` is roughly the same performance as a forward loop. See updated thread.

File: extensions/amp-image-slider/0.1/amp-image-slider.js (hunk 30 above)
Comment (kristoferbaxter) [reply extensions/amp-image-slider/0.1/amp-image-slider.js:51]: This convention is not equally followed within the file, going to leave this one be.

File: extensions/amp-image-slider/0.1/amp-image-slider.js (hunk 32 above)
Comment (kristoferbaxter) [reply extensions/amp-image-slider/0.1/amp-image-slider.js:304]: Good call, resolving.

File: builtins/amp-img/amp-img.js [hunk 33]
```diff
@@ -130,8 +131,9 @@ export class AmpImg extends BaseElement {
           this.element
         );
       }
-      this.propagateAttributes(
+      propagateAttributes(
```
Comment (alanorozco) [inline builtins/amp-img/amp-img.js:134]: Unrelated to this change: I wonder how much we could save from compiling out these `BaseElement` helpers everywhere.

File: extensions/amp-image-slider/0.1/test/integration/test-amp-image-slider.js [hunk 34]
```diff
@@ -26,17 +26,14 @@ t.run('amp-image-slider', {}, function () {
   // #s2 has attribute `disable-hint-reappear` set
   // A huge padding is added to the bottom to allow room for scrolling
   const sliderBody = `
```
Comment (alanorozco) [inline extensions/amp-image-slider/0.1/test/integration/test-amp-image-slider.js:28]: nit: you get prettier formatting if you add a leading comment.

```suggestion
  const sliderBody = /* HTML */ `
```

**Review by alanorozco — APPROVED** [review]

File: builtins/amp-img/amp-img.js (hunk 33 above)
Comment (alanorozco) [reply builtins/amp-img/amp-img.js:134]: After looking further, it looks like there's only two.

File: extensions/amp-image-slider/0.1/test/integration/test-amp-image-slider.js (hunk 34 above)
Comment (kristoferbaxter) [reply extensions/amp-image-slider/0.1/test/integration/test-amp-image-slider.js:28]: Oooh, that's a nice tip!

**Review by rcebulko — APPROVED** [review]
Bundle size
